storage/innobase/buf/buf0flu.cc

```diff
@@ -2087,6 +2091,83 @@ bool buf_flush_do_batch(buf_pool_t *buf_pool, buf_flush_t type, ulint min_n,
   return (true);
 }
 
+/** Waits until a flush batch of the given lsn ends
+@param[in]	new_oldest	target oldest_modified_lsn to wait for
+                                if LSN_MAX, we wait for flush of all
+                                non-temporary tablespace pages to be
+                                flushed */
+static void buf_flush_wait_flushed(lsn_t new_oldest) {
+  lsn_t instance_newest[MAX_BUFFER_POOLS];
+  for (ulint i = 0; i < srv_buf_pool_instances; ++i) {
+    instance_newest[i] = 0;
+
+    buf_pool_t *buf_pool{buf_pool_from_array(i)};
+
+    if (new_oldest == LSN_MAX) {
+      buf_flush_list_mutex_enter(buf_pool);
+
+      buf_page_t *bpage{UT_LIST_GET_FIRST(buf_pool->flush_list)};
+
+      while (bpage != nullptr && fsp_is_system_temporary(bpage->id.space())) {
+        bpage = UT_LIST_GET_NEXT(list, bpage);
+      }
+
+      if (bpage != nullptr) {
+        ut_ad(bpage->in_flush_list);
+        instance_newest[i] = bpage->oldest_modification;
+      }
+
+      buf_flush_list_mutex_exit(buf_pool);
+    }
+  }
+
+  for (ulint i = 0; i < srv_buf_pool_instances; ++i) {
+    lsn_t oldest;
+    buf_pool_t *buf_pool{buf_pool_from_array(i)};
+
+    for (;;) {
+      /* We don't need to wait for fsync of the flushed
+      blocks, because anyway we need fsync to make chekpoint.
+      So, we don't need to wait for the batch end here. */
+
+      buf_flush_list_mutex_enter(buf_pool);
+
+      buf_page_t *bpage{nullptr};
+
+      /* We don't need to wait for system temporary pages */
+      for (bpage = UT_LIST_GET_LAST(buf_pool->flush_list);
```

Comment:
Have you considered just waiting for buf_pool->flush_list to become empty ?

Reply:
Yes, it seems we are trying to optimize by not waiting for temp tablespace pages to be flushed.

So, by doing that (just waiting for flush_list to become empty), we will remove optimization and make it worse(may be)

Reply:
yes, makes sense.